src/libmongoc/src/mongoc/mongoc-util.c

```diff
@@ -338,6 +338,27 @@ _mongoc_validate_replace (const bson_t *doc,
       return false;
    }
 
+   if (!bson_iter_init (&iter, doc)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_BSON,
+                      MONGOC_ERROR_BSON_INVALID,
+                      "replace document is corrupt");
+      return false;
+   }
+
+   while (bson_iter_next (&iter)) {
+      key = bson_iter_key (&iter);
+      if (key[0] == '$') {
+         bson_set_error (error,
+                         MONGOC_ERROR_COMMAND,
+                         MONGOC_ERROR_COMMAND_INVALID_ARG,
+                         "Invalid key '%s': replace prohibits $ operators",
+                         key);
```

Comment:
This is basically the inverse of the existing logic in `_mongoc_validate_update`. We technically only need to check the first key since the server would raise an error (as [discussed in the CRUD spec](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#update-vs-replace-validation)), but I opted to keep this the same as `update` validation. If anyone feels strongly about this, feel free to open a new CDRIVER ticket to propose a future change to both validation methods.